karrot/activities/api.py

```diff
@@ -252,3 +254,22 @@ def remove(self, request, pk=None):
     def conversation(self, request, pk=None):
         """Get conversation ID of this activity"""
         return self.retrieve_conversation(request, pk)
+
+
+class ActivityICSViewSet(mixins.RetrieveModelMixin, GenericViewSet):
+    serializer_class = ActivityICSSerializer
+    renderer_classes = (ICSCalendarRenderer, )
+    queryset = ActivityModel.objects
+    filter_backends = (filters.DjangoFilterBackend, )
+    filterset_class = ActivitiesFilter
+    permission_classes = (IsAuthenticated, )
+
+    def get_queryset(self):
+        return self.queryset.filter(place__group__members=self.request.user)
+
+    def finalize_response(self, request, response, *args, **kwargs):
+        """specify a filename for the downloaded file"""
+        super(ActivityICSViewSet, self).finalize_response(request, response, *args, **kwargs)
+        if isinstance(response, Response) and response.accepted_renderer.format == 'ics':
+            response['content-disposition'] = 'attachment; filename=invite.ics'
+        return response
```

Comment:
Instead of a new viewset (which then has it's own url prefix `/activities-ics/`), I think this could be neater as a detail method on the existing viewset, something like:

```py
    @action(
        detail=True,
        methods=['GET'],
        renderer_classes=(ICSCalendarRenderer, ),
        serializer_class=ActivityICSSerializer,
    )
    def ics(self, request, pk=None):
        response = self.retrieve(request)
        response['content-disposition'] = 'attachment; filename=invite.ics'
        return response
```

What do you think? 

(I'd also suggest making the filename unique to the activity, in the simplest way `activity-{id}.ics` perhaps? Or `{type}-{id}.ics`?)

Reply:
sure why not. with a viewset i was thinking you could then also expose ics files for all pickups in a place, in a group, joined by a user… so not just a 'retrieve' view, but also a list view with filters etc. cause ics files can contain many events! but its not supported atm and maybe we can migrate to a viewset later if need be? i guess its no big deal if urls change

Reply:
Yes, I thought about that too, it _looks like_ you can also do these `@action` things with `detail=False` and it would create an endpoint like `/activities/ics/` that comes 1 higher priority than `/activities/{id}/` so I _think_ that would work... didn't try it though.

I like the idea though, and it could be used with https://icsx5.bitfire.at/ nicely. For that subscription case though we'd need to think about the authentication aspect too... maybe it'd need a seperate viewset for that, or whatever. We can think of that if someone is motivated to build it :)